File: backend/app/api/features.py
"""Feature request endpoints."""

import uuid
from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import desc, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.db import get_db
from backend.app.models.feature import FeatureRequest, FeatureVote
from backend.app.models.user import User
from backend.app.schemas.feature import FeatureCreate, FeatureResponse, FeatureVoteCreate
from backend.app.services.broadcaster import broadcast_event, feature_update_event

router = APIRouter(prefix="/features", tags=["features"])


@router.get("", response_model=list[FeatureResponse])
async def list_features(
    status: str | None = None,
    db: AsyncSession = Depends(get_db),
) -> list[dict]:
    query = (
        select(
            FeatureRequest,
            func.coalesce(func.sum(FeatureVote.vote), 0).label("vote_count"),
        )
        .outerjoin(FeatureVote, FeatureRequest.id == FeatureVote.feature_id)
        .group_by(FeatureRequest.id)
    )

    if status:
        query = query.where(FeatureRequest.status == status)

    query = query.order_by(desc(FeatureRequest.created_at))
    result = await db.execute(query)
    rows = result.all()

    return [
        {
            "id": fr.id,
            "title": fr.title,
            "description": fr.description,
            "status": fr.status,
            "created_by": fr.created_by,
            "created_at": fr.created_at,
            "vote_count": vote_count,
        }
        for fr, vote_count in rows
    ]


@router.post("", response_model=FeatureResponse, status_code=201)
async def create_feature(data: FeatureCreate, db: AsyncSession = Depends(get_db)) -> dict:
    user_result = await db.execute(select(User).where(User.type == "human"))
    user = user_result.scalar_one_or_none()
    if not user:
        raise HTTPException(status_code=400, detail="No user onboarded")

    feature = FeatureRequest(
        id=str(uuid.uuid4()),
        title=data.title,
        description=data.description,
        status="open",
        created_by=user.id,
        created_at=datetime.now(UTC).isoformat(),
    )
    db.add(feature)
    await db.flush()

    # Broadcast new feature creation
    await broadcast_event(
        feature_update_event(
            feature_id=feature.id,
            title=feature.title,
            status=feature.status,
            vote_count=0,
            update_type="created",
        )
    )

    return {
        "id": feature.id,
        "title": feature.title,
        "description": feature.description,
        "status": feature.status,
        "created_by": feature.created_by,
        "created_at": feature.created_at,
        "vote_count": 0,
    }


@router.post("/{feature_id}/vote", status_code=200)
async def vote_feature(
    feature_id: str,
    data: FeatureVoteCreate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    # Verify feature exists
    fr = await db.execute(select(FeatureRequest).where(FeatureRequest.id == feature_id))
    if not fr.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Feature not found")

    user_result = await db.execute(select(User).where(User.type == "human"))
    user = user_result.scalar_one_or_none()
    if not user:
        raise HTTPException(status_code=400, detail="No user onboarded")

    if data.vote not in (1, -1):
        raise HTTPException(status_code=400, detail="Vote must be +1 or -1")

    # Upsert vote
    existing = await db.execute(
        select(FeatureVote).where(
            FeatureVote.feature_id == feature_id,
            FeatureVote.user_id == user.id,
        )
    )
    vote = existing.scalar_one_or_none()
    if vote:
        vote.vote = data.vote
    else:
        db.add(FeatureVote(feature_id=feature_id, user_id=user.id, vote=data.vote))
    await db.flush()

    # Get feature title and updated vote count for broadcast
    feature_result = await db.execute(select(FeatureRequest).where(FeatureRequest.id == feature_id))
    feature = feature_result.scalar_one()
    count_result = await db.execute(
        select(func.coalesce(func.sum(FeatureVote.vote), 0)).where(
            FeatureVote.feature_id == feature_id
        )
    )
    vote_count = count_result.scalar()

    await broadcast_event(
        feature_update_event(
            feature_id=feature_id,
            title=feature.title,
            status=feature.status,
            vote_count=vote_count,
            update_type="voted",
        )
    )

    return {"status": "voted", "vote": data.vote}

File: server/src/routes/features.ts
/**
 * Feature request endpoints.
 */

import { Hono } from "hono";
import { eq, desc, sql } from "drizzle-orm";
import { getDb } from "../db";
import { featureRequests, featureVotes, users } from "../db/schema";
import { FeatureCreateSchema, FeatureVoteCreateSchema } from "../types";
import { broadcastEvent, featureUpdateEvent } from "../services/broadcaster";
import type { FeatureResponse } from "../types";

const app = new Hono();

// GET /features
app.get("/", (c) => {
  const status = c.req.query("status");
  const db = getDb();

  let query = db
    .select({
      id: featureRequests.id,
      title: featureRequests.title,
      description: featureRequests.description,
      status: featureRequests.status,
      createdBy: featureRequests.createdBy,
      createdAt: featureRequests.createdAt,
      voteCount: sql<number>`coalesce(sum(${featureVotes.vote}), 0)`,
    })
    .from(featureRequests)
    .leftJoin(featureVotes, eq(featureRequests.id, featureVotes.featureId))
    .groupBy(featureRequests.id)
    .$dynamic();

  if (status) {
    query = query.where(eq(featureRequests.status, status));
  }

  const rows = query.orderBy(desc(featureRequests.createdAt)).all();

  const result: FeatureResponse[] = rows.map((r) => ({
    id: r.id,
    title: r.title,
    description: r.description,
    status: r.status,
    created_by: r.createdBy,
    created_at: r.createdAt,
    vote_count: Number(r.voteCount),
  }));

  return c.json(result);
});

// POST /features
app.post("/", async (c) => {
  const body = await c.req.json();
  const parsed = FeatureCreateSchema.safeParse(body);
  if (!parsed.success) {
    return c.json({ detail: parsed.error.message }, 400);
  }
  const db = getDb();

  const human = db.select().from(users).where(eq(users.type, "human")).get();
  if (!human) {
    return c.json({ detail: "No user onboarded" }, 400);
  }

  const id = crypto.randomUUID();
  const now = new Date().toISOString();

  db.insert(featureRequests)
    .values({
      id,
      title: parsed.data.title,
      description: parsed.data.description,
      status: "open",
      createdBy: human.id,
      createdAt: now,
    })
    .run();

  broadcastEvent(
    featureUpdateEvent({
      featureId: id,
      title: parsed.data.title,
      status: "open",
      voteCount: 0,
      updateType: "created",
    })
  );

  return c.json(
    {
      id,
      title: parsed.data.title,
      description: parsed.data.description,
      status: "open",
      created_by: human.id,
      created_at: now,
      vote_count: 0,
    } satisfies FeatureResponse,
    201
  );
});

// POST /features/:featureId/vote
app.post("/:featureId/vote", async (c) => {
  const featureId = c.req.param("featureId");
  const body = await c.req.json();
  const parsed = FeatureVoteCreateSchema.safeParse(body);
  if (!parsed.success) {
    return c.json({ detail: parsed.error.message }, 400);
  }
  const db = getDb();

  // Verify feature exists
  const feature = db
    .select()
    .from(featureRequests)
    .where(eq(featureRequests.id, featureId))
    .get();
  if (!feature) {
    return c.json({ detail: "Feature not found" }, 404);
  }

  const human = db.select().from(users).where(eq(users.type, "human")).get();
  if (!human) {
    return c.json({ detail: "No user onboarded" }, 400);
  }

  // Upsert vote
  const existing = db
    .select()
    .from(featureVotes)
    .where(
      sql`${featureVotes.featureId} = ${featureId} AND ${featureVotes.userId} = ${human.id}`
    )
    .get();

  if (existing) {
    db.update(featureVotes)
      .set({ vote: parsed.data.vote })
      .where(
        sql`${featureVotes.featureId} = ${featureId} AND ${featureVotes.userId} = ${human.id}`
      )
      .run();
  } else {
    db.insert(featureVotes)
      .values({
        featureId,
        userId: human.id,
        vote: parsed.data.vote,
      })
      .run();
  }

  // Get updated vote count
  const countRow = db
    .select({
      total: sql<number>`coalesce(sum(${featureVotes.vote}), 0)`,
    })
    .from(featureVotes)
    .where(eq(featureVotes.featureId, featureId))
    .get();
  const voteCount = Number(countRow?.total ?? 0);

  broadcastEvent(
    featureUpdateEvent({
      featureId,
      title: feature.title,
      status: feature.status,
      voteCount,
      updateType: "voted",
    })
  );

  return c.json({ status: "voted", vote: parsed.data.vote });
});

export default app;
